packages/block-editor/src/store/selectors.js

```diff
@@ -137,32 +140,76 @@ export function getBlockAttributes( state, clientId ) {
+ *    block tree, but stop when it reaches a different entity, like a template
+ *    part. In this scenario, withImmediateChildren is false deep into the block
+ *    tree. In other words, if an InnerBlock controller is reached, the selector
+ *    is only memoized based on the controller's attributes, NOT the block cache
+ *    state. The block cache state changes when the controlled inner blocks
+ *    change, which is not applicable to a parent entity, so we do not want to
+ *    trigger any updates.
+ * 2. A child controller. For the child controller, it needs to be aware of its
+ *    direct children. As a result, it is memoized on the block cache state,
+ *    since it does want to be aware of changes to its inner blocks. If it were
+ *    only memoized on attributes, it would not update as its inner blocks
+ *    change.
+ *
+ * In most cases, both of the above are true. For example, a template part acts
+ * as a "child" controller since it updates as its inner blocks change. But as
+ * it traverses into the block tree, it acts as a "parent", since
+ * withImmediateChildren will not be passed down as this selector recurses. So,
+ * as a controller, it can listen directly as its inner blocks change without
+ * seeing any changes inside of any controllers which may happen to exist in its
+ * inner blocks.
  *
+ * @param {Object}  state    Editor state.
+ * @param {string}  clientId Block client ID.
+ * @param {boolean} [withImmediateChildren] Use this parameter to ignore whether
+ *                                          inner blocks are controlled. Helpful
+ *                                          for the inner block component, which
+ *                                          needs to see its children, even though
+ *                                          they are controlled.
  * @return {Object} Parsed block object.
  */
 export const getBlock = createSelector(
-	( state, clientId ) => {
+	( state, clientId, withImmediateChildren = false ) => {
 		const block = state.blocks.byClientId[ clientId ];
 		if ( ! block ) {
 			return null;
 		}
 
+		const innerBlocks =
+			! withImmediateChildren &&
+			areInnerBlocksControlled( state, clientId )
+				? EMPTY_ARRAY
+				: getBlocks( state, clientId );
 		return {
 			...block,
+			innerBlocks,
 			attributes: getBlockAttributes( state, clientId ),
-			innerBlocks: getBlocks( state, clientId ),
 		};
 	},
-	( state, clientId ) => [
+	( state, clientId, withImmediateChildren = false ) => {
+		// From the perspective of most entities, we do not want to refresh the
+		// selector when a different entity's inner blocks change. So in the case
+		// of controlled inner blocks, only use block attributes. If withImmediateChildren
+		// is set, then we do care about changes to child blocks.
+		if (
+			! withImmediateChildren &&
+			areInnerBlocksControlled( state, clientId )
+		) {
+			return [ getBlockAttributes( state, clientId ) ];
```

Comment:
This is buggy for me because controller blocks doesn't necessarily  mean immediate children. So if I update a paragraph  block inside a group block, I'm still inside the same entity, so the group block should refresh to include the changes of the paragraph block. 

i also believe we should only update the "cache" key in the reducer for performance reasons and  not  try to do the magic here. I mean the "cache"   key should be updated when  the block needs to be updated (taking into consideration that changes to inner blocks from controlled inner blocks shouldn't change its parents  cache key)

Reply:
If you need help with the reducer logic, I can try to chime-in if you want.

Reply:
>So if I update a paragraph block inside a group block, I'm still inside the same entity, so the group block should refresh to include the changes of the paragraph block.

That still works. Group blocks are not controlled inner blocks.

>I mean the "cache" key should be updated when the block needs to be updated (taking into consideration that changes to inner blocks from controlled inner blocks shouldn't change its parents cache key)

That won't work for the root block list. We would need to introduce a root cache key for `getBlocks`.

Reply:
> That still works. Group blocks are not controlled inner blocks.

Ok, now I understand the logic better I think this still needs to be done on the reducer based on the cache key for performance reasons.

 > That won't work for the root block list. We would need to introduce a root cache key for getBlocks.

I'm still having trouble understanding this, I'm proposing the same logic to be moved to the cache key, we're returning the cache key of the "getBlock" here and not the "getBlocks", so it's still a per block cache key.

Reply:
@youknowriad it needs to be both for the logic I mentioned in the above comment. There are two separate separate situations where we get the blocks which need to be cached differently

From the perspective of an InnerBlock controller, when a child block is updated, we need the cache key to change on the child block and the parent inner block controller. If the cache key of the parent inner block controller does not change, then it will not see the change to its block that it knows about. We do not want any cache keys to update for parents of the inner block controller, and we have cache logic in this PR to prevent that from happening.

So that works. If a controlled inner block is updated, its cache key change and so does the cache key of its parent, the controller. And that's all controlled by the existing cache state.

However, this is problematic because from the perspective of the higher level entity, if it is _only_ based on the same cache key, then it will think that the controlled inner block has changed since the cache key of the controller has changed! So that won't work for us, because it will make a higher level entity think that it has pending changes.

So when we are looking at the inner block controller from the perspective of a higher-level entity, what do we care about? We only care about the attributes, since attributes are stored in the level of the higher-level entity. Again, if this higher level entity sees the cache key of the controller changing (as noted above), then it will think that it has pending changes. But the conundrum is if we stop changing the cache key of the controller, then the controller won't know about it's own changes! So we do need to cache it differently from the different perspectives. As it stands, this attribute logic is only used in _one_ situation: when a parent entity has recursed into the block tree and selects an inner block controller. That selector on the inner block controller will only change when the attributes change, plus it will not return the inner blocks. 

Every other scenario uses the normal cache key logic in the state.

(The only other solution I could think of was creating an entirely new cache state.)

Reply:
If a top-level template part has child inner block changes and it changes its cache key, `getBlocks` would return a different array.

Conversely, if we don't update the cache key, the `getBlock` in `InnerBlocks` will never update, and we won't see the changes.

Essentially, what Noah did here holds two cache keys for the same block. One used by the parent and one used by it's `InnerBlocks`. This could be done in the cache reducer, but I assume it was a lot more complicated.

Reply:
Also, I want to highlight again that the only time the `getBlock` selector is cached on block attributes is if (and only if) a top level entity has recursed into the block tree and sees an inner block _controller_ in the state. I think the number of blocks (all template parts right now) cached on that is very low, and I think the attributes change very infrequently for template parts.

Every other `getBlock` selector is cached on the normal cache state.

So consider a very simple template that has three template parts in this shape:

```
template:
  tpA:
    tpChild:
  otherBlocks
  tpB:
```

`getBlock` will be cached on attributes only three times:
1. `getBlocks` on the template has recursed and called `getBlock( tpA )`
2. `getBlock` on tpA has recursed and called `getBlock( tpChild )`
3. `getBlocks` on the template has recursed and called `getBlock( tpB )`

Plus, take case two as an example. The only time we want `tpA` to become dirty on a change to `tpChild` is if `tpChild`'s direct attributes change (e.g. the template part ID). So caching on attributes in this scenario accounts for that.

Reply:
> but I assume it was a lot more complicated.

Yeah, I'm not sure how we can do it in the cache reducer without changing a lot about how it works, because it's impossible to cache a block on the same key from the perspective of both things.

Reply:
We would need two keys using a prefix or something.

Reply:
At the time, it seemed like the update logic would be hard (but I think knowing that only block attributes need to be detected, it wouldn't be that hard). I still think the current implication is fine -- I guess we would have to change the logic to reference `cache[ something + clientID ]` instead of `getBlockAttributes`. I'm not convinced that really improves anything

Reply:
Not performance-wise, but it is cleaner.

Reply:
I'm going to comment as I read.

> From the perspective of an InnerBlock controller, when a child block is updated, we need the cache key to change on the child block and the parent inner block controller. If the cache key of the parent inner block controller does not change, then it will not see the change to its block that it knows about

For  me in this situation   we're not asking for the "block" object, we're just asking directly for the inner block, it seems that this is a need for the `getBlocks( rootClientId )` and not `getBlock` for me

Reply:
`InnerBlocks` has always used `getBlock( ... ).innerBlocks` because `getBlocks` doesn't have a cache.

Reply:
I'm having trouble following for me here  are the expected results:

```
getBlock( clientId ) // Returns the block object which is clientId  including its inner blocks but the hierarchy of  innerBlocks stops the moment we find a controlled container block,  we include it in the hierarchy but not its blocks

getBlocks( clientId ) // Returns the inner  blocks of a given (clientId), if clientId is null,  it will return the root level blocks. also for each block in the returned array, it's j just computed using the regular getBlock as explained above.
```

Where do you think this falls short?

Reply:
The array returned by `getBlocks` is not cached. See the definition.

Reply:
> Returns the block object which is clientId including its inner blocks but the hierarchy of innerBlocks stops the moment we find a controlled container block,  we include it in the hierarchy but not its blocks

> Returns the inner  blocks of a given (clientId), if clientId is null,  it will return the root level blocks. also for each block in the returned array, it's just computed using the regular getBlock as explained above

This is exactly how it works with this implementation :) 

I think what's unclear is why we need the variable. If you think about it, when we use `getBlocks` we call `getBlock` on every inner block. So we call `getBlock( templatePartId )` for a template part. In this scenario, we do NOT want any inner blocks returned. How can we differentiate that situation from your first paragraph? Here, we don't want the inner blocks. But when we call `getBlock( templatePartId )` as you say in your first paragraph, we do want to include inner blocks.

The current implementation uses the variable to differentiate between those two cases

Reply:
I think what @youknowriad is suggesting is that a simpler alternative to maintaining this sort of dual cache mechanism would be:

- Cache keys never change for controller blocks when their children change.
- `InnerBlocks` uses `getBlocks( rootClientId )` so that it can still rerender changes.
- `getBlocks`'s cache key becomes its shallow list of `InnerBlocks`.

If I'm not missing anything, this would be a lot simpler.

Reply:
Definitely simpler to understand :)

Reply:
It should now use `getBlocks`